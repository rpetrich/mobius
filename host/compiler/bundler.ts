import Concat from "concat-with-sourcemaps";
import { Finaliser, OutputOptions, Plugin, RollupCache } from "rollup";
import _rollupBabel from "rollup-plugin-babel";
import { RawSourceMap } from "source-map";
import { packageRelative } from "../fileUtils";
import memoize from "../memoize";
import { ModuleMap } from "../modules/index";
import { staticFileRoute, StaticFileRoute } from "../static-file-route";
import { Compiler } from "./compiler";

export interface CompiledRoute {
	route: StaticFileRoute;
	map?: RawSourceMap;
}

export interface CompilerOutput {
	routes: { [path: string]: CompiledRoute };
	moduleMap: ModuleMap;
}

const declarationOrJavaScriptPattern = /\.(d\.ts|js)$/;

const requireOnce = memoize(require);

type PluginReference = [string | Function, any] | [string | Function];

function lazilyLoadedBabelPlugins(references: PluginReference[]): any[] {
	const result: any[] = [];
	for (let i = 0; i < references.length; i++) {
		const ref = references[i];
		const name = ref[0];
		if (typeof name === "string") {
			Object.defineProperty(result, i, {
				configurable: true,
				get() {
					const plugin = requireOnce(name);
					const value = [plugin.default || plugin, ref[1] || {}];
					Object.defineProperty(result, i, {
						configurable: true,
						value,
					});
					return value;
				},
			});
		} else {
			result[i] = ref;
		}
	}
	result.length = references.length;
	return result;
}

export type CacheData = RollupCache;

type DependencyData = [string, string] & ReadonlyArray<string | number>;

export async function bundle(compiler: Compiler<CacheData>, appPath: string, publicPath: string, minify: boolean, redact: boolean, fileRead: (path: string) => void): Promise<CompilerOutput> {
	// Dynamically load dependencies to reduce startup time
	const rollupModule = await import("rollup");
	const rollupBabel = requireOnce("rollup-plugin-babel") as typeof _rollupBabel;

	const compiled = compiler.compile();
	const rollupPre: Plugin = {
		name: "mobius-pre",
		resolveId(importee: string, importer: string | undefined) {
			// Windows
			if (importer) {
				importer = importer.replace(/\\/g, "/");
			}
			try {
				const result = compiled.resolveModule(importee, importer !== undefined ? importer : compiler.basePath);
				if (result) {
					return result.resolvedFileName;
				}
			} catch (e) {
				if (!e || e.code !== "MODULE_NOT_FOUND") {
					throw e;
				}
			}
		},
		load(id: string) {
			const module = compiled.getVirtualModule(id);
			if (module) {
				return module.generateModule();
			}
		},
		transform(code, id) {
			const output = compiled.getEmitOutput(id.toString());
			if (output) {
				let code = output.code;
				const mappingIndex = code.lastIndexOf("//# sourceMap" + "pingURL=");
				if (mappingIndex !== -1) {
					const newlineIndex = code.indexOf("\n", mappingIndex);
					if (newlineIndex !== -1) {
						code = code.substring(0, mappingIndex) + code.substring(newlineIndex);
					} else {
						code = code.substring(0, mappingIndex);
					}
				}
				return {
					code,
					map: typeof output.map === "string" ? JSON.parse(output.map) : undefined,
				};
			}
		},
	};

	// Transform the intermediary phases via babel
	const babelPlugin = rollupBabel({
		babelrc: false,
		presets: [],
		plugins: lazilyLoadedBabelPlugins(([
			["babel-plugin-syntax-dynamic-import"],
			["babel-plugin-external-helpers"],
			["babel-plugin-transform-async-to-promises", { externalHelpers: true, hoist: true, minify: true }],
			["babel-plugin-optimize-closures-in-render"],
		] as PluginReference[]).concat(redact ? [["./stripRedactedArguments"]] as PluginReference[] : []).concat([
			["./fixTypeScriptExtendsWarning"],
			["./noImpureGetters"],
			["./simplifyVoidInitializedVariables"],
			["./stripUnusedArgumentCopies"],
			// Replacement for babel-preset-env
			["babel-plugin-check-es2015-constants"],
			["babel-plugin-syntax-trailing-function-commas"],
			["babel-plugin-transform-es2015-arrow-functions"],
			["babel-plugin-transform-es2015-block-scoped-functions"],
			["babel-plugin-transform-es2015-block-scoping"],
			["babel-plugin-transform-es2015-classes"],
			["babel-plugin-transform-es2015-computed-properties"],
			["babel-plugin-transform-es2015-destructuring"],
			["babel-plugin-transform-es2015-duplicate-keys"],
			["babel-plugin-transform-es2015-for-of"],
			["babel-plugin-transform-es2015-function-name"],
			["babel-plugin-transform-es2015-literals"],
			["babel-plugin-transform-es2015-object-super"],
			["babel-plugin-transform-es2015-parameters"],
			["babel-plugin-transform-es2015-shorthand-properties"],
			["babel-plugin-transform-es2015-spread"],
			["babel-plugin-transform-es2015-sticky-regex"],
			["babel-plugin-transform-es2015-template-literals"],
			["babel-plugin-transform-es2015-typeof-symbol"],
			["babel-plugin-transform-es2015-unicode-regex"],
			["babel-plugin-transform-exponentiation-operator"],
			["./simplifyTypeof"],
		])),
	});

	const plugins = [rollupPre, babelPlugin];
	// If minifying, use Closure Compiler
	if (minify) {
		plugins.push(requireOnce("rollup-plugin-closure-compiler-js")({
			languageIn: "ES5",
			languageOut: "ES3",
			assumeFunctionWrapper: false,
			rewritePolyfills: false,
		}) as Plugin);
	}

	let remainingOutputCount = 0;
	let resolveOutput: () => void | undefined;
	const waitForOutput = new Promise((resolve) => resolveOutput = resolve);
	plugins.push({
		name: "mobius-post",
		transform(code, id) {
			// Track input files read so the --watch option works
			fileRead(id.toString());
		},
		ongenerate(options, chunk) {
			const path = chunk.map!.file;
			routes[path] = {
				route: staticFileRoute("/" + (minify && path != mainChunkId ? routeIndexes.indexOf(path).toString(36) + ".js" : path), chunk.code),
				map: chunk.map!,
			};
			if (--remainingOutputCount === 0) {
				resolveOutput!();
			}
		},
	});

	const mainChunkId = "main.js";
	const routes: { [path: string]: CompiledRoute } = {};
	const moduleMap: ModuleMap = {};
	const routeIndexes: string[] = [];
	const moduleDependencies: { [name: string]: Array<string | number> } = {};
	const routeReferenceForId = (id: string) => {
		let index = routeIndexes.indexOf(id);
		if (index === -1) {
			index = routeIndexes.length;
			routeIndexes.push(id);
		}
		return minify ? index : id;
	};
	const customFinalizer: Finaliser = {
		name: minify ? "mobius-minified" : "mobius",
		supportsCodeSplitting: true,
		async finalise(
			magicString,
			{
				id,
				dependencies,
				modules,
				exports,
				namedExportsMode,
				generateExportBlock,
			},
			options: OutputOptions,
		) {
			dependencies = dependencies.slice();
			const isMain = id === mainChunkId;
			if (isMain) {
				// Coordinate with output of other finalizers
				await Promise.resolve();
				if (remainingOutputCount === 0) {
					resolveOutput!();
				}
				await waitForOutput;
			} else {
				if (routeIndexes.indexOf(id) === -1) {
					routeIndexes.push(id);
				}
				remainingOutputCount++;
			}

			// Bundle any CSS provided by the modules in the chunk (only virtual modules can provide CSS)
			const cssModuleName = id.replace(/(\.js)?$/, ".css");
			const css = new Concat(true, cssModuleName, minify ? "" : "\n\n");
			const bundledCssModulePaths: string[] = [];
			for (const moduleId of Object.keys(modules)) {
				const virtualModule = compiled.getVirtualModule(moduleId);
				if (virtualModule && virtualModule.generateStyles) {
					bundledCssModulePaths.push(moduleId);
					const styles = virtualModule.generateStyles(modules[moduleId].renderedExports);
					if (styles.css) {
						css.add(moduleId, styles.css, styles.map);
					}
				}
			}

			// Register CSS route
			let cssRoute: StaticFileRoute | undefined;
			const cssString = css.content.toString();
			if (cssString) {
				const mapString = css.sourceMap;
				const cssMap = mapString ? JSON.parse(mapString) : undefined;
				cssRoute = staticFileRoute("/" + cssModuleName, cssString);
				if (!isMain && routeIndexes.indexOf(cssModuleName) === -1) {
					routeIndexes.push(cssModuleName);
				}
				for (const bundledModuleName of bundledCssModulePaths) {
					moduleMap[bundledModuleName.replace(declarationOrJavaScriptPattern, "")] = cssRoute.foreverPath;
				}
				routes[cssModuleName] = {
					route: cssRoute,
					map: cssMap,
				};
			}

			// Generate code to ask for and receive imported modules
			const mainIndex = dependencies.findIndex((m) => m.id === "./" + mainChunkId);
			let mainIdentifier: string = "main";
			if (mainIndex !== -1) {
				mainIdentifier = dependencies[mainIndex].name;
				dependencies.splice(mainIndex, 1);
			}
			const deps = dependencies.map((m) => m.id.substring(2)).concat(cssRoute && !isMain ? [cssModuleName] : []).map(routeReferenceForId);
			moduleDependencies[id] = deps;
			const args = dependencies.map((m) => m.name);
			if (args.length || mainIndex !== -1) {
				args.unshift(mainIdentifier);
			}
			args.unshift("_import");

			// Generate code to write exported symbols into the exports object
			args.unshift("exports");
			const exportBlock = generateExportBlock();
			if (exportBlock) {
				magicString.append("\n\n" + exportBlock, {});
			}
			magicString.append("\n}", {});

			if (isMain) {
				args.push("document");
				if (cssRoute) {
					// Coordinate load with the main.css so that we don't inadvertently mutate the DOM before it's ready
					magicString.prepend(
						`var i=0,` +
						`stylesheets=document.querySelectorAll("link"),` +
						`link=document.createElement("link");` +
						`link.href=${JSON.stringify(cssRoute.foreverPath)};` +
						`if("onload" in link){` +
							`for(_mobius=link.onload=loaded;i<stylesheets.length;i++)` +
								`if(stylesheets[i].href==link.href)` +
									`return stylesheets[i].sheet ? loaded() : stylesheets[i].onload=loaded;` +
						`}else ` +
							`main();` +
						`link.rel="stylesheet";` +
						`link.setAttribute("integrity",${JSON.stringify(cssRoute.integrity)});` +
						`document.head.appendChild(link);` +
						`function loaded(){` +
							`main();` +
							`if(link=document.querySelector("style#mobius-inlined"))` +
								`link.parentNode.removeChild(link)` +
						`}` +
						`function main() {\n`);
				} else {
					magicString.prepend(`\n`);
				}
				// Add JavaScript equivalent of frame-ancestors 'none'
				magicString.prepend(
					`if (top != self) {` +
						`document.open();` +
						`document.close();` +
						`return;` +
					`}`);
				// Check that property iteration matches V8's behavior and that there aren't additional properties on Array or Object's prototype
				magicString.prepend(
					`var unspecifiedBehaviorCheck = [0], unspecifiedBehaviorOrder = [{ a: 0, 0: 0, length: 0 }], k;` +
					`unspecifiedBehaviorCheck["1"] = unspecifiedBehaviorCheck["a"] = 0;` +
					`for (k in unspecifiedBehaviorCheck) ` +
						`unspecifiedBehaviorOrder.push(k);` +
					`for (k in unspecifiedBehaviorOrder[0]) ` +
						`unspecifiedBehaviorOrder.push(k);` +
					`unspecifiedBehaviorOrder.push(unspecifiedBehaviorCheck.length);` +
					`if (JSON.stringify(unspecifiedBehaviorOrder) != '[{"0":0,"a":0,"length":0},"0","1","a","0","a","length",2]') ` +
						`return;`);
				// Add sanity check for prerequisites, will early exit to fallback
				magicString.prepend(
					`if (!Object.keys) ` +
						`return;`);
				// Add IIFE wrapper
				magicString.prepend(`(function(${args.join(", ")}) {`);
				function loadDataForModuleWithName(name: string): DependencyData {
					const route = routes[name].route;
					const result: DependencyData = [route.foreverPath.substr(1), route.integrity];
					if (Object.hasOwnProperty.call(moduleDependencies, name)) {
						return result.concat(moduleDependencies[name]) as DependencyData;
					}
					return result;
				}
				// Insert imports mapping, using an array and indexes when minified
				let imports: DependencyData[] | { [name: string]: DependencyData };
				if (minify) {
					imports = routeIndexes.map(loadDataForModuleWithName);
				} else {
					const importsObject: { [path: string]: string[] } = imports = {};
					routeIndexes.forEach((path) => importsObject[path] = loadDataForModuleWithName(path));
				}
				if (cssRoute) {
					magicString.append("}");
				}
				magicString.append(`)({}, ${JSON.stringify(imports)}, document)`);
			} else {
				// Generate code to inform the loader that our module's content has loaded
				magicString.prepend(`_mobius(function(${args.join(", ")}) {\n`);
				const argumentJSON = JSON.stringify([minify ? routeIndexes.indexOf(id) : id].concat(deps));
				magicString.append(", " + argumentJSON.substr(1, argumentJSON.length - 2) + ")");
			}

			return magicString;
		},
		finaliseDynamicImport(magicString, { importRange, argumentRange }) {
			magicString.overwrite(importRange.start, argumentRange.start, `_import(`);
			magicString.overwrite(argumentRange.end, importRange.end, `)`);
		},
		dynamicImportArgument(path) {
			return JSON.stringify(routeReferenceForId(path.substring(2)));
		},
		reservedIdentifiers: ["_import"],
	};
	const bundle = await rollupModule.rollup({
		input: [packageRelative("common/main.ts")],
		cache: compiler.cache.data,
		external(id: string, parentId: string, isResolved: boolean) {
			return false;
		},
		plugins,
		acorn: {
			allowReturnOutsideFunction: true,
		},
		// Use experimental rollup features, including the aggressive merging features our fork features
		experimentalCodeSplitting: true,
		experimentalPreserveModules: false,
		aggressivelyMergeIntoEntryPoint: true,
		inlineDynamicImports: false,
	});
	const cacheSave = compiled.saveCache(bundle.cache);
	// Generate the output, using our custom finalizer for client
	const output = (await bundle.generate({
		format: customFinalizer,
		sourcemap: true,
		name: "app",
		compact: minify,
	})).output;
	// Fill in source maps
	for (const id of Object.keys(output)) {
		const chunk = output[id];
		if (typeof chunk !== "string" && !(chunk instanceof Buffer)) {
			routes[id].map = chunk.map;
		}
	}
	await cacheSave;
	return {
		routes,
		moduleMap,
	};
}
