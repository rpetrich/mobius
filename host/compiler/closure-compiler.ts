export default function closure() {
	return {
		name: "closure-compiler",
		transformBundle(src: string) {
			const { compile } = require("google-closure-compiler-js");
			const output = compile({
				languageIn: "ES5",
				languageOut: "ES5",
				assumeFunctionWrapper: false,
				rewritePolyfills: false,
				createSourceMap: true,
				processCommonJsModules: true,
				jsCode: [{ src }],
			});
			return {
				code: output.compiledCode,
				map: output.sourceMap,
			};
		},
	};
}
