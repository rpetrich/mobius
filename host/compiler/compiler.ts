import { readFileSync } from "fs";
import { resolve } from "path";
import { cwd } from "process";
import * as ts from "typescript";
import { exists, mkdir, modifiedTime, packageRelative, readJSON, writeFile } from "../fileUtils";
import { typescript } from "../lazy-modules";
import { virtualModule } from "../lazy-modules";
import { once } from "../memoize";
import { VirtualModule } from "../modules/index";

function loadCompilerOptions(mode: "server" | "client", mainPath: string) {
	const basePath = resolve(mainPath, "..");
	const fileName = `tsconfig-${mode}.json`;
	const configFile = typescript.readJsonConfigFile(packageRelative(fileName), (path: string) => readFileSync(path).toString());
	const configObject = typescript.convertToObject(configFile, []);
	const result = typescript.convertCompilerOptionsFromJson(configObject.compilerOptions, packageRelative("./"), fileName).options;
	result.suppressOutputPathCheck = true;
	result.baseUrl = basePath;
	result.paths = {
		"app": [
			mainPath,
		],
		"*": [
			packageRelative(`${mode}/*`),
			resolve(basePath, `${mode}/*`),
			packageRelative("common/*"),
			resolve(basePath, "common/*"),
			packageRelative("types/*"),
			resolve(basePath, "*"),
		],
		"tslib": [
			packageRelative("node_modules/tslib/tslib"),
		],
		"babel-plugin-transform-async-to-promises/helpers": [
			resolve(require.resolve("babel-plugin-transform-async-to-promises"), "../helpers"),
		],
		"preact": [
			packageRelative("dist/common/preact"),
		],
		"redom": [
			packageRelative("host/redom"),
		],
	};
	return result;
}

const diagnosticsHost = {
	getCurrentDirectory: cwd,
	getCanonicalFileName(fileName: string) {
		return fileName;
	},
	getNewLine() {
		return "\n";
	},
};

const declarationPattern = /\.d\.ts$/;

const getDocumentRegistry = once(() => typescript.createDocumentRegistry());

export interface CompiledOutput<T> {
	program: ts.Program;
	compiler: Compiler<T>;
	resolveModule(moduleName: string, containingFile: string): { resolvedFileName: string, isExternalLibraryImport?: boolean } | void;
	getAllFileNames(): string[];
	getEmitOutput(path: string): { code: string, map: string | undefined } | void;
	getVirtualModule(path: string): VirtualModule | void;
	saveCache(newCache: T): Promise<void>;
}

const compilerModifiedTime = once(() => modifiedTime(packageRelative("dist/host/compiler/compiler.js")));

interface VirtualModuleCacheEntry {
	dependencies: { [dependency: string]: number };
	declaration?: string;
	code?: string;
}

export interface CacheData<T> {
	path: string;
	data?: T;
	virtualModules: { [module: string]: VirtualModuleCacheEntry };
}

const version: number = 1;

export async function loadCache<T>(mainPath: string, cacheSlot: string): Promise<CacheData<T>> {
	const path = resolve(mainPath, `../.cache/${cacheSlot}.json`);
	if (await exists(path) && (modifiedTime(path) > compilerModifiedTime())) {
		const result = await readJSON(path);
		if (result && result.version === version) {
			return {
				path,
				data: result.data,
				virtualModules: result.virtualModules || { },
			};
		}
	}
	return {
		path,
		virtualModules: { },
	};
}

export function noCache<T>(): CacheData<T> {
	return {
		path: "",
		virtualModules: { },
	};
}

function notExternal<T extends ts.ResolvedModule | undefined>(resolvedModule: T): T {
	// Override to allow paths inside node_modules/mobius-js/ to be emitted by TypeScript
	if (resolvedModule && !/\.d\.ts$/.test(resolvedModule.resolvedFileName)) {
		resolvedModule.isExternalLibraryImport = false;
	}
	return resolvedModule;
}

export class Compiler<T> {
	public readonly basePath: string;
	public readonly compilerOptions: ts.CompilerOptions;
	private readonly paths: string[];
	private readonly host: ts.LanguageServiceHost & ts.ModuleResolutionHost;
	private readonly languageService: ts.LanguageService;
	private readonly resolutionCache: ts.ModuleResolutionCache;
	private virtualModules: { [path: string]: VirtualModule } = { };
	private virtualModuleEntries: { [module: string]: VirtualModuleCacheEntry } = { };

	constructor(mode: "server" | "client", public readonly cache: CacheData<T>, mainPath: string, rootFileNames: string[], private minify: boolean, private fileRead: (path: string) => void) {
		this.basePath = resolve(mainPath, "..");
		this.compilerOptions = loadCompilerOptions(mode, mainPath);
		this.paths = (require("module")._nodeModulePaths(this.basePath) as string[]).concat(module.paths);
		const readFile = (path: string, encoding?: string) => {
			const module = this.getVirtualModule(path);
			if (module) {
				return module.generateTypeDeclaration();
			}
			fileRead(path);
			return typescript.sys.readFile(path, encoding);
		};
		this.resolutionCache = typescript.createModuleResolutionCache(this.basePath, (s) => s);
		this.host = {
			getScriptFileNames() {
				return rootFileNames;
			},
			getScriptVersion: (fileName) => {
				return modifiedTime(fileName).toString();
			},
			getScriptSnapshot(fileName) {
				const contents = readFile(fileName);
				if (typeof contents !== "undefined") {
					return typescript.ScriptSnapshot.fromString(contents);
				}
				return undefined;
			},
			getCurrentDirectory() {
				return typescript.sys.getCurrentDirectory();
			},
			getCompilationSettings: () => {
				return this.compilerOptions;
			},
			getDefaultLibFileName(options) {
				return typescript.getDefaultLibFilePath(options);
			},
			readFile,
			fileExists: (path: string) => {
				const result = typescript.sys.fileExists(path);
				if (result) {
					return result;
				}
				if (this.getVirtualModule(path)) {
					return true;
				}
				return false;
			},
			readDirectory: typescript.sys.readDirectory,
			directoryExists(directoryName: string): boolean {
				return typescript.sys.directoryExists(directoryName);
			},
			getDirectories(directoryName: string): string[] {
				return typescript.sys.getDirectories(directoryName);
			},
			resolveModuleNames: (moduleNames: string[], containingFile: string, reusedNames?: string[]) => {
				return moduleNames.map((moduleName) => notExternal(typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache).resolvedModule!));
			},
			getResolvedModuleWithFailedLookupLocationsFromCache: (moduleName: string, containingFile: string) => {
				const result = typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache);
				notExternal(result.resolvedModule);
				return result;
			},
		};
		this.languageService = typescript.createLanguageService(this.host, getDocumentRegistry());
	}

	public fileChanged = (path: string) => {
		/* tslint:disable no-empty */
	}

	private getVirtualModule = (path: string) => {
		if (declarationPattern.test(path)) {
			path = path.replace(declarationPattern, "");
			if (Object.hasOwnProperty.call(this.virtualModules, path)) {
				return this.virtualModules[path];
			}
			const entry: VirtualModuleCacheEntry = {
				dependencies: { },
			};
			const result = virtualModule.default(this.basePath, path, this.minify, (filePath: string) => {
				entry.dependencies[filePath] = modifiedTime(filePath);
				this.fileRead(path);
			}, this.compilerOptions);
			if (result) {
				this.virtualModuleEntries[path] = entry;
				if (this.cache && this.cache.virtualModules && Object.hasOwnProperty.call(this.cache.virtualModules, path)) {
					const pendingCacheEntry = this.cache.virtualModules[path];
					let stale = false;
					for (const dependency in pendingCacheEntry.dependencies) {
						if (Object.hasOwnProperty.call(pendingCacheEntry.dependencies, dependency)) {
							if (modifiedTime(dependency) !== pendingCacheEntry.dependencies[dependency]) {
								stale = true;
								break;
							}
						}
					}
					if (!stale) {
						entry.dependencies = Object.assign({}, pendingCacheEntry.dependencies);
						if (typeof pendingCacheEntry.declaration !== "undefined") {
							entry.declaration = pendingCacheEntry.declaration;
						}
						if (typeof pendingCacheEntry.code !== "undefined") {
							entry.code = pendingCacheEntry.code;
						}
					}
				}
				// Declaration is eager
				const declaration = entry.declaration || (entry.declaration = result.generateTypeDeclaration());
				return this.virtualModules[path] = {
					generateTypeDeclaration() {
						return declaration;
					},
					generateModule() {
						// Module is lazy
						return entry.code || (entry.code = result.generateModule());
					},
					instantiateModule: result.instantiateModule,
					generateStyles: result.generateStyles,
				};
			}
		}
	}

	public compile(): CompiledOutput<T> {
		this.virtualModules = { };
		const program = this.languageService.getProgram();
		if (typeof program === "undefined") {
			throw new Error("Expected language service to have a program!");
		}
		const diagnostics = typescript.getPreEmitDiagnostics(program);
		if (diagnostics.length) {
			console.log(typescript.formatDiagnostics(diagnostics, diagnosticsHost));
		}
		return {
			program,
			compiler: this,
			resolveModule: (moduleName: string, containingFile: string): { resolvedFileName: string, isExternalLibraryImport?: boolean } | void => {
				const tsResult = typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache).resolvedModule;
				if (tsResult && !tsResult.isExternalLibraryImport) {
					if (tsResult.extension === ".d.ts") {
						const replaced = tsResult.resolvedFileName.replace(declarationPattern, ".js");
						if (this.host.fileExists(replaced)) {
							return {
								resolvedFileName: replaced,
								isExternalLibraryImport: tsResult.isExternalLibraryImport,
							};
						}
					}
					return {
						resolvedFileName: tsResult.resolvedFileName,
						isExternalLibraryImport: tsResult.isExternalLibraryImport,
					};
				}
				return {
					resolvedFileName: (module.constructor as any)._resolveFilename(moduleName, null, false, { paths: this.paths }) as string,
					isExternalLibraryImport: true,
				};
			},
			getAllFileNames() {
				return program.getSourceFiles().map(value => value.fileName);
			},
			getEmitOutput: (path: string): { code: string, map: string | undefined } | void => {
				const sourceFile = program.getSourceFile(path);
				if (sourceFile) {
					let code: string | undefined;
					let map: string | undefined;
					for (const { name, text } of this.languageService.getEmitOutput(path).outputFiles) {
						if (/\.js$/.test(name)) {
							code = text;
						} else if (/\.js\.map$/.test(name)) {
							map = text;
						}
					}
					if (typeof code === "string") {
						return { code, map };
					}
				}
			},
			getVirtualModule: this.getVirtualModule,
			saveCache: async (data: T) => {
				this.cache.data = data;
				this.cache.virtualModules = this.virtualModuleEntries;
				this.virtualModuleEntries = { };
				const path = this.cache.path;
				if (path) {
					const parentPath = resolve(path, "..");
					if (!await exists(parentPath)) {
						await mkdir(parentPath);
					}
					await writeFile(path, JSON.stringify({ version, data, virtualModules: this.cache.virtualModules }));
				}
			},
		};
	}

}
