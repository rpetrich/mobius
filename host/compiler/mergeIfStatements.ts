import { NodePath } from "babel-traverse";
import { IfStatement, isBooleanLiteral, isIfStatement, isReturnStatement, logicalExpression, Node } from "babel-types";

function isReturnFalseIfStatement(node: Node): node is IfStatement {
	return isIfStatement(node) && !node.alternate && isReturnStatement(node.consequent) && isBooleanLiteral(node.consequent.argument) && !node.consequent.argument.value;
}

export default function() {
	return {
		visitor: {
			IfStatement(path: NodePath<IfStatement>) {
				const container = path.container;
				if (typeof (container as any).length == "number" && isReturnFalseIfStatement(path.node)) {
					const index = (container as any[]).indexOf(path.node);
					if (index > 0) {
						const previous = path.getSibling((index - 1) as any as string);
						if (previous.isIfStatement() && isReturnFalseIfStatement(previous.node)) {
							previous.get("test").replaceWith(logicalExpression("||", previous.node.test, path.node.test));
							path.remove();
						}
					}
				}
			},
		},
	};
}
