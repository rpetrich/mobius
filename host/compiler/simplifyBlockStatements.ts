import { NodePath } from "babel-traverse";
import { BlockStatement } from "babel-types";

export default function() {
	return {
		visitor: {
			BlockStatement(path: NodePath<BlockStatement>) {
				if (typeof (path.container as any).length == "number" && path.node.body.length) {
					path.replaceWithMultiple(path.node.body);
				}
			},
		},
	};
}
