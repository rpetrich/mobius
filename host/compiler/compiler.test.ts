import { noCache, Compiler } from "./compiler";
import { packageRelative } from "../fileUtils";

const randomness = packageRelative("tests/randomness/app.tsx");

describe("Compiler", () => {

	describe("server", () => {
		const compiler = new Compiler("server", noCache(), randomness, [packageRelative("server/dom-ambient.d.ts"), packageRelative("common/main.ts")], false, (path: string) => {
		});
		const compiled = compiler.compile();
		it("randomness/app.tsx output", () => {
			const output = compiled.getEmitOutput(randomness) as { code: string }
			expect(output.code).toMatchSnapshot();
		});
	});

	describe("client", () => {
		const compiler = new Compiler("client", noCache(), randomness, [packageRelative("common/main.ts")], false, (path: string) => {
		});
		const compiled = compiler.compile();
		it("randomness/app.tsx output", () => {
			const output = compiled.getEmitOutput(randomness) as { code: string }
			expect(output.code).toMatchSnapshot();
		});
	});

});
