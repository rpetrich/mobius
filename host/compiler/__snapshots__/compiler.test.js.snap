// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler client randomness/app.tsx output 1`] = `
"import * as dom from \\"dom\\";
for (let i = 0; i < 50; i++) {
    console.log(Math.random());
}
export default dom.h(\\"div\\", null);
//# sourceMappingURL=app.js.map"
`;

exports[`Compiler server randomness/app.tsx output 1`] = `
"import * as dom from \\"dom\\";
for (let i = 0; i < 50; i++) {
    console.log(Math.random());
}
export default dom.h(\\"div\\", null);
//# sourceMappingURL=app.js.map"
`;
